File: go/stack/stack.go
package main

import "fmt"

type Node struct {
	value int
	prev  *Node
}

var head *Node = nil
var length int = 0

func push(value int) {
	var newNode *Node = &Node{value: value}
	length += 1

	if head == nil {
		head = newNode
	} else {
		newNode.prev = head
		head = newNode
	}
}

func pop() int {
	if head == nil {
		return -1
	}
	length = max(0, length-1)

	if length == 0 {
		var temp *Node = head
		value := temp.value
		head = nil
		return value
	}

	var temp *Node = head
	value := temp.value
	head = head.prev
	return value
}

func peek() int {
	return head.value
}

func main() {
	push(5)
	push(3)
	push(7)
	push(9)
	pop()
	fmt.Println(head.value)
	fmt.Println(length)
}

File: c_c++/stack/stack.c
#include <stdio.h>
#include <stdlib.h>
#define MAX(x, y) (((x) > (y)) ? (x) : (y))

struct Node
{
    int value;
    struct Node *prev;
};

struct Node *head = NULL;
int length = 0;

void push(int value)
{
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    newNode->value = value;
    newNode->prev = NULL;
    length++;
    if (head == NULL)
    {
        head = newNode;
    }
    else
    {
        newNode->prev = head;
        head = newNode;
    }
}

int pop()
{
    if (head == NULL)
        return -1;

    length = MAX(0, length - 1);

    if (length == 0)
    {
        struct Node *temp = head;
        int value = temp->value;
        head = NULL;
        free(temp);
        return value;
    }

    struct Node *temp = head;
    int value = temp->value;
    head = head->prev;
    free(temp);

    return value;
}

int peek()
{
    return head->value;
};

int main()
{
    push(5);
    push(2);
    push(6);
    push(10);
    pop();
    printf("%d\n", head->value);
    printf("%d\n", length);
}